### PR TITLE
[20.01] Retry reading tool conf on IOError

### PR DESCRIPTION
**Pull request by mvdbeek**
I think that might stabilize the test_tool_reload* unit tests.

### CODE REVIEW

File: lib/galaxy/tools/toolbox/watcher.py [hunk 1]
```diff
@@ -131,11 +133,16 @@ def check(self):
                     # in rare cases `path` may be deleted between `os.path.exists` calls
                     # and reading the file from the filesystem. We do not want the watcher
                     # thread to die in these cases.
-                    try:
-                        del hashes[path]
-                        del paths[path]
-                    except KeyError:
-                        pass
+                    if path in self.drop_now:
+                        log.warning("'%s' could not be read, removing from watched files")
```
Comment (nsoranzo) [inline lib/galaxy/tools/toolbox/watcher.py]:
```suggestion
                        log.warning("'%s' could not be read, removing from watched files", path)
```

File: lib/galaxy/tools/toolbox/watcher.py [hunk 2]
```diff
@@ -71,6 +71,8 @@ class ToolConfWatcher(object):
 
     def __init__(self, reload_callback, tool_cache=None):
         self.paths = {}
+        self.drop_on_next_loop = set()
+        self.drop_now = set()
```
Comment (nsoranzo) [inline lib/galaxy/tools/toolbox/watcher.py]: Shouldn't `drop_on_next_loop` and `drop_now` be local variables inside the `check()` method (like `hashes`)?

Comment (mvdbeek) [reply lib/galaxy/tools/toolbox/watcher.py]: Yes, good thinking!

File: lib/galaxy/tools/toolbox/watcher.py [hunk 3]
```diff
@@ -131,11 +133,16 @@ def check(self):
...
-                        del paths[path]
-                    except KeyError:
-                        pass
+                    if path in drop_now:
+                        log.warning("'%s' could not be read, removing from watched files", path)
+                        try:
+                            del hashes[path]
+                            del paths[path]
+                        except KeyError:
+                            pass
```
Comment (nsoranzo) [inline lib/galaxy/tools/toolbox/watcher.py]: If `path` is not in `hashes`, `path` would never be removed from `paths`.

```suggestion
                        del paths[path]
                        if path in hashes:
                            del hashes[path]
```

File: lib/galaxy/tools/toolbox/watcher.py [hunk 4]
```diff
@@ -131,11 +133,14 @@ def check(self):
...
-                        del hashes[path]
+                    if path in drop_now:
+                        log.warning("'%s' could not be read, removing from watched files", path)
                         del paths[path]
-                    except KeyError:
-                        pass
+                        if path in hashes:
+                            del hashes[path]
+                    else:
+                        log.debug("'%s could not be read")
```
Comment (nsoranzo) [inline lib/galaxy/tools/toolbox/watcher.py]:
```suggestion
                        log.debug("'%s could not be read", path)
```

**Review by nsoranzo — APPROVED** [review]

**Comment by nsoranzo** [comment]
@mvdbeek I've retargeted this to 20.01 (I think that was your original intention, given the branch history?) and fixed the `test_run_simple` test in py27-unit also for 20.01. Hope that's OK.

**Comment by mvdbeek** [comment]
sure, thanks for that. let's hope we see a bit more green soon!

**Pull request merged**